### PR TITLE
resource(engine): Improve RPC method error handling

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: ref #6208

### What is changed and how it works?
- Adopted `engine/pkg/rpcerror` to standardize gRPC methods' return errors.
- Added appropriate retry mechanisms in the `engine/pkg/client` package.

#### Tests 

 - Unit test
 - Integration test

##### Will it cause performance regression or break compatibility?
No.

##### Do you need to update user documentation, design documentation or monitoring documentation?
No.

### Release note 

```release-note
None.
```

### CODE REVIEW

File: engine/pkg/client/resource_manager_client.go [hunk 1]
```diff
@@ -37,21 +39,69 @@ func NewResourceManagerClient(cli enginepb.ResourceManagerClient) ResourceManage
...
+	})
+	if err != nil {
+		return err
+	}
+
+	// Check whether the previous request has actually succeeded.
+	if queryResp.CreatorWorkerId == request.CreatorWorkerId {
+		return nil
+	}
+	return originalErr
```
Comment (amyangfei) [inline engine/pkg/client/resource_manager_client.go:78]: Does this happen when two workers create the same resource?

File: engine/pkg/externalresource/internal/errors.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	ResourceID resModel.ResourceID
+	Details    string
+}
+
+// ErrResourceMetastoreError indicates that there is a failure in querying the
+// metastore on a resource, but the error can be non-fatal.
+var ErrResourceMetastoreError = rpcerror.Normalize[ResourceMetastoreError]()
+
+// ResourceAlreadyExistsErrorInfo provides details of ErrResourceAlreadyExists
+type ResourceAlreadyExistsErrorInfo struct {
```
Comment (gozssky) [inline engine/pkg/externalresource/internal/errors.go:46]: This name with the `Info` suffix is inconsistent with other places, like `ResourceMetastoreError`, and errors defined in https://github.com/pingcap/tiflow/blob/master/engine/servermaster/scheduler/errors.go.

File: engine/pkg/externalresource/manager/service.go [hunk 3]
```diff
@@ -43,46 +42,65 @@ type Service struct {
...
 	shouldRet, err := s.preRPCHook.PreRPC(ctx, request, &resp2)
 	if shouldRet {
 		return resp2, err
 	}
 
-	record, err := s.metaclient.GetResourceByID(ctx,
-		pkgOrm.ResourceKey{JobID: request.GetResourceKey().GetJobId(), ID: request.GetResourceKey().GetResourceId()})
+	defer func() {
+		if retErr != nil {
+			retErr = rpcerror.ToGRPCError(retErr)
```
Comment (gozssky) [inline engine/pkg/externalresource/manager/service.go:77]: https://github.com/pingcap/tiflow/pull/6669 will add a server interceptor to convert error. So we don't need to convert the error manually.

File: engine/pkg/client/resource_manager_client.go (hunk 1 above)
Comment (liuzix) [reply engine/pkg/client/resource_manager_client.go:78]: No. It can only happen if the same worker has tried to create the same resource twice.

**Review by amyangfei — APPROVED** [review]

**Review by gozssky — APPROVED** [review]

**Pull request merged**